Issue 22099

sujalshah-bit opened the pull request:
<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sujalshah-bit commented:
We currently have blue links for the 'check all' and 'uncheck all' buttons in the multiuse modal, like this:
![image](https://github.com/zulip/zulip/assets/73663475/ed7e0fd1-0bfe-4715-96bf-7a28725844af)

What I think is, we should change it to look like this:
![image](https://github.com/zulip/zulip/assets/73663475/28d1327d-aa75-4651-87f7-5835d5eff101)

Additionally, we should use email pills and user pills here:
![image](https://github.com/zulip/zulip/assets/73663475/f0a2c6ef-8b78-40b9-bd9d-c3892aed8c90)

@alya  your thoughts on this and kindly review PR :)

timabbott commented:
@Vector73 might be worth you reviewing this since you've been touching the API documentation for this endpoint.

Vector73 reviewed (commented):
@sujalshah-bit Thanks for working on this. I have left a few comments.

alya commented:
> Additionally, we should use email pills and user pills here:

That sounds great!

alya commented:
For "Check all" / "Uncheck all" , I think that will become moot after #26871.

sujalshah-bit commented on zerver/views/invite.py:
```diff
@@ -185,6 +186,38 @@ def resend_user_invite_email(
+        raise JsonableError(
+            _("Invite does not exist with id: {invite_id}.").format(invite_id=invite_id)
+        )
+
+    check_role_based_permissions(invite_as, user_profile, require_admin=True)
+
```
non admins don't have permission to create multiuse invite link
![image](https://github.com/zulip/zulip/assets/73663475/b1c5a34c-bc75-497c-8eaf-65907db52e5c)

Vector73 replied:
You have to change the "Who can create reusable invitation links" permission in  "Organization settings > Organization permissions" as administrator to allow non-admin users to create invite links.

sujalshah-bit commented:
Sorry for the late response due to my exams. @Vector73, I've made the changes according to your feedback
and for 
>This logic will only allow admins to edit the multiuse invite.

left a comment above 

can you kindly review :)

Vector73 commented on zerver/actions/invites.py:
```diff
@@ -399,11 +399,26 @@ def do_get_invites_controlled_by_user(user_profile: UserProfile) -> List[Dict[st
                 ),
                 invited_as=invite.invited_as,
                 is_multiuse=True,
+                stream_ids=list(invite.streams.values_list("id", flat=True)),
```
It looks kind of odd to me to only return `stream_ids` for multi-use invite links. I think it would make sense to return them for email invitations as well and the API changelog should be updated accordingly.

timabbott replied:
Yeah, agreed.

Vector73 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,13 @@ format used by the Zulip server that they are interacting with.
 ## Changes in Zulip 9.0
 
+**Feature level 254**
+
+* [`GET /invites`](/api/get-invites): Added the `stream_ids` field to be returned for all
+* `PATCH /invites/multiuse/{invite_id}`: Added support for editing multiuse invites
+  to users as admin with regards to role and associated streams.
+  multiuse invite objects.
```
Some lines are messed up here.

timabbott commented on zerver/views/invite.py:
```diff
@@ -185,6 +186,38 @@ def resend_user_invite_email(
+        streams.append(stream)
+
+    do_edit_multiuse_invite_link(invite, invite_as, streams)
+    return json_success(request)
```
This seems to do only about half the security checks from `generate_multiuse_invite_backend`. The result is that your code likely introduces at least one security bug. Please be careful about this sort of thing.

timabbott commented on web/src/settings_invites.ts:
```diff
@@ -148,7 +148,7 @@ function do_revoke_invite({
     const modal_invite_id = $(".dialog_submit_button").attr("data-invite-id");
     const modal_is_multiuse = $(".dialog_submit_button").attr("data-is-multiuse");
-    const $revoke_button = $row.find("button.revoke");
+    const $revoke_button = $row.find("button.revoke-invite");
```
This button change commit could be usefully pulled out as its own preparatory PR.

timabbott commented on web/src/invite.ts:
```diff
@@ -351,6 +352,11 @@ function open_invite_user_modal(e: JQuery.ClickEvent<Document, undefined>): void
+        const default_stream = streams.filter((stream) => stream.default_stream);
+        for (const stream of default_stream) {
+            $(`input[value=${stream.stream_id}]`).prop("checked", true);
+        }
```
Why are these JS changes required for a commit that appears to mainly just refactor some template code?

sujalshah-bit commented:
### Changes I have made following changes:


**Major changes:**
<details>
  <summary>Documented PATCH `/invites/multiuse/{invite_id}`</summary>

![localhost_9991_api_edit-invite-link](https://github.com/zulip/zulip/assets/73663475/f95b3986-bdc4-4f5b-8571-17c8b4537696)

  
</details>

<details>
  <summary>Created help center article for edit multiuse  invites</summary>

- I think it make sense to document this features.

![help_edit_resusable_guide](https://github.com/zulip/zulip/assets/73663475/69101c02-e0b0-490c-aac3-dbace6b9b1d4)

  
</details>

<details>
  <summary>Documented DELETE `/invites/multiuse/{invite_id}`</summary>
  <br>

-  I have documented this endpoint because CI checks were failing. This is due to I had documented the same endpoint for the `PATCH` method but not for the `DELETE` method. Additionally, this endpoint was listed as a `pending_endpoint`, so I documented it in a separate commit.


![localhost_9991_api_revoke-reusable-invite](https://github.com/zulip/zulip/assets/73663475/7c121859-03b5-4217-92d3-a93184205691)



</details>

-----

**Minor changes:**
- Updated the `Changes` section of the `/invites:get` endpoint to document the change in `stream_ids`.
  
- Added a separate commit for button change in its own preparatory PR.
  
- Changed the commit message for the 6th commit regarding the form partial.
  
- Added security checks for `edit_multiuse_invite` and tests related to the added checks.
  
- Added `stream_ids` for email invitations as well and updated the changelog accordingly.

@timabbott  @Vector73 kindly review :)

sujalshah-bit commented:
The changelog.md file is constantly updating, can I add its content at the end when the PR is ready to merge. This way, I won't have to resolve conflicts again and again.

Vector73 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,14 @@ format used by the Zulip server that they are interacting with.
 ## Changes in Zulip 9.0
 
+**Feature level 257**
+
+* [`GET /invites`](/api/get-invites): The stream_ids field has been added to
+  be returned for all invites.
```
You should change stream_ids  to \`stream_ids\` here.

Vector73 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,14 @@ format used by the Zulip server that they are interacting with.
+
+* [`GET /invites`](/api/get-invites): The stream_ids field has been added to
+  be returned for all invites.
+* [`PATCH /invites/multiuse/{invite_id}`](/api/edit-invite-link): Added support for
+  editing multiuse invites to users as admin or non admin with regards to role and
+  associated streams.
+
```
The wording doesn't seem clear to me. Maybe we can change it to something like: 
```
Added support for editing `invite_as` and `stream_ids` fields of reusable invitation links.
```
@timabbott thoughts?

Vector73 commented on zerver/openapi/zulip.yaml:
```diff
@@ -12180,6 +12185,156 @@ paths:
+      summary: Edit a reusable invitation link
+      tags: ["invites"]
+      description: |
+        [Edit a reusable invitation link](/help/invite-new-users#edit-a-reusable-invitation-link)
+        which can be used to invite new users to the organization.
```
I think you should mention the fields that can be changed by using this endpoint.

Vector73 commented:
I think you forgot to address these comments: [#discussion_r1573464597](https://github.com/zulip/zulip/pull/29664#discussion_r1573464597), [#discussion_r1573464712](https://github.com/zulip/zulip/pull/29664#discussion_r1573464712).

whilstsomebody commented:
Hey @sujalshah-bit , can I continue with your work on this issue?
